physics/cires_ugwp_module_v1.F90

```diff
@@ -0,0 +1,666 @@
+!
+! call init-solvers for "broad" non-stationary multi-wave spectra
+!
+    if   (knob_ugwp_solver==1) then
+!
+      call initsolv_lsatdis(me, master, knob_ugwp_wvspec(2), knob_ugwp_azdir(2), &
+                            knob_ugwp_stoch(2), knob_ugwp_effac(2), do_physb_gwsrcs, kxw )
+    endif
+     if   (knob_ugwp_solver==2) then 
+!
+! re-assign from namelists 
+!     
+       nslope = knob_ugwp_nslope             ! the GW sprctral slope at small-m             
+       lzstar = knob_ugwp_lzstar
+       lzmax  = knob_ugwp_lzmax 
+       lzmin  = knob_ugwp_lzmin      
+       lhmet  = knob_ugwp_lhmet
+       tamp_mpa =knob_ugwp_tauamp           !amplitude for GEOS-5/MERRA-2
+       tau_min  =knob_ugwp_taumin             ! min of GW MF 0.25 mPa                        
+       ilaunch = launch_level
+       kxw = pi2/lhmet
+       call initsolv_wmsdis(me, master, knob_ugwp_wvspec(2), knob_ugwp_azdir(2), &
+                   knob_ugwp_stoch(2), knob_ugwp_effac(2), do_physb_gwsrcs, kxw)
+     endif
+!
+! other solvers not yet tested for fv3gfs
+!
+!<    if   (knob_ugwp_solver==3) call init_dspdis
+!<    if   (knob_ugwp_solver==4) call init_adodis
+! 
+
+!======================
+    module_is_initialized = .true.
+    if (me == master)   print *, ' CIRES-ugwp-V1 is initialized ', module_is_initialized
+
+    end subroutine cires_ugwp_init_v1
+
+    
+!=============================================
+
+
+     subroutine cires_ugwp_advance
+!-----------------------------------------------------------------------
+!
+!   options for the day-to-day variable sources/spectra + diagnostics
+!           for stochastic "triggers"
+!   diagnose GW-source functions * FGF + OKWP + SGO/CONV from IAU-fields
+!     or use for stochastic GWP-sources "memory"
+!-----------------------------------------------------------------------
+      implicit none
+!
+! update sources
+!  a) physics-based triggers for multi-wave
+!  b) stochastic-based spectra and amplitudes
+!  c) use "memory" on GW-spectra from previous time-step
+!  d) update "background" GW  dissipation as needed 
+!
+     end subroutine cires_ugwp_advance
+ 
+!      
+! -----------------------------------------------------------------------
+! finalize  of cires_ugwp   (_finalize)
+! -----------------------------------------------------------------------
+
+
+  subroutine cires_ugwp_finalize
+!
+! deallocate sources/spectra & some diagnostics need to find where "deaalocate them"
+! before "end" of the FV3GFS
+!
+    implicit none
+!
+!   deallocate arrays employed in:
+!     cires_ugwp_advance / cires_ugwp_driver / cires_ugwp_init
+!
+    deallocate( kvg,   ktg  )
+    deallocate( krad,  kion )
+    deallocate( zkm,   pmb  )
+    deallocate( rfdis, rfdist)
+    deallocate(ugwp_taulat,  ugwp_qbolat)
+    deallocate(tau_limb, uzmf_merra)
+    deallocate(days_limb,  days_merra,  pmb127)
+ 
+   end subroutine cires_ugwp_finalize
+
+!
+!
+!
+!
+    subroutine calendar_ugwp(yr, mm, dd, ddd_ugwp)
```

Comment:
The Julian day is available from the host model, no need to calculate again. From `GFS_typdefs.meta`
```
[julian]
  standard_name = julian_day
  long_name = julian day
  units = days
  dimensions = ()
  type = real
  kind = kind_phys
```
Just need to make sure when the call to `calendar_ugwp` occurs in your current code. In the `init` phase? If so, we may have to add the julian day calculation to the `init` phase of `GFS_phys_time_vary_init` subroutine.

Reply:
Is it okay if we delay this change?  I had spoken about this issue with Valery Yudin last month, and there's a reason he is calculating the dates/times this way.  I would like to get him involved with this change, but he's not available for another couple of weeks.

Reply:
Fine with me. We need to create an issue with follow-up changes to the unified GWD scheme that contains everything we didn't address in this PR.